File: frontend/src/types/session.ts
import type { ReportOutput } from "./report";

export interface ScenarioSummary {
  scenario_code: string;
  title: string;
  background: string;
  estimated_minutes: number;
  version: string;
  source_type: "seeded" | "seeded_fallback" | "ai_base" | "ai_adapted" | string;
}

export const OCCUPATION_CATEGORIES = [
  "学生",
  "教育培训",
  "医疗健康",
  "互联网/信息技术",
  "工程/制造/建筑",
  "商业/金融/管理",
  "政府/公共服务",
  "科研/法律/专业服务",
  "文化/传媒/创意",
  "零售/餐饮/生活服务",
  "自由职业/个体经营",
  "待业/退休/其他",
] as const;

export type OccupationCategory = (typeof OCCUPATION_CATEGORIES)[number];

export interface StageSummary {
  stage_code: string;
  title: string;
  stage_order: number;
  context: string;
  main_question: string;
  max_followups: number;
}

export interface DialogueTurnItem {
  turn_index: number;
  speaker: "ai" | "user" | "system" | string;
  content: string;
  content_type: string;
  created_at: string;
  analysis?: Record<string, unknown> | null;
}

export interface StageProgressItem {
  stage_code: string;
  title: string;
  stage_order: number;
  status: "pending" | "active" | "completed" | string;
  max_followups: number;
  used_followups: number;
  used_clarifications: number;
  can_skip: boolean;
  skipped: boolean;
  released_dynamic_info_count: number;
  estimated_minutes: number;
  evidence_coverage?: Record<string, "complete" | "partial" | "missing" | string>;
  missing_evidence?: string[];
  waiting_for_stage_choice?: boolean;
}

export interface SessionProgress {
  total_stages: number;
  current_stage_order: number | null;
  estimated_minutes: number;
  elapsed_seconds: number | null;
  stages: StageProgressItem[];
}

export interface SessionResponse {
  session_uuid: string;
  status: string;
  flow_version?: "legacy_v2" | "progressive_v3" | "progressive_v3_2" | "progressive_v3_3";
  interviewer_style_version?: string;
  participant_nickname: string;
  scenario: ScenarioSummary;
  current_stage: StageSummary | null;
  turns: DialogueTurnItem[];
  progress?: SessionProgress | null;
  interview_progress?: {
    formal_answer_count: number;
    target_min_answers: number;
    target_max_answers: number;
    percent: number;
    estimated_remaining_minutes: number;
    elapsed_seconds?: number;
  } | null;
  language_mode: "standard" | "plain";
  phase: "onboarding" | "scenario_preparing" | "opening_pending" | "assessment" | "completed";
  onboarding?: {
    question_count: number;
    max_questions: number;
    completed: boolean;
  } | null;
  scenario_preparation?: {
    status: string;
    cache_hit: boolean;
    fallback_used: boolean;
    message?: string | null;
  } | null;
}

export interface CreateSessionPayload {
  nickname: string;
  occupation_category: OccupationCategory;
  occupation: string;
  info_collect_method?: string;
  assessment_mode?: string;
  consent_accepted: true;
  consent_version: "critical_thinking_assessment_consent_v1";
}

export interface PreparationResponse {
  session_uuid: string;
  phase: SessionResponse["phase"];
  onboarding: NonNullable<SessionResponse["onboarding"]>;
  scenario_preparation: NonNullable<SessionResponse["scenario_preparation"]>;
  assessment_ready: boolean;
}

export interface SubmitTurnPayload {
  content: string;
  content_type?: string;
  client_turn_id: string;
  answer_duration_ms?: number | null;
}

export interface SubmitTurnResponse {
  session_uuid: string;
  saved_turn_index: number;
  next_action: string;
  message: string;
  replayed?: boolean;
}

export interface FinishSessionResponse {
  session_uuid: string;
  status: string;
  completed_at: string;
}

export interface SkipStageResponse {
  session_uuid: string;
  next_action: string;
  message: string;
}

export interface ContinueStageResponse extends SkipStageResponse {}

export interface LanguageModeResponse {
  session_uuid: string;
  language_mode: "standard" | "plain";
}

export interface SubmitFeedbackPayload {
  realism_score: number;
  difficulty_score: number;
  naturalness_score: number;
  fatigue_score: number;
  report_trust_score: number;
  overall_satisfaction_score: number;
  open_feedback?: string | null;
}

export interface FeedbackResponse extends SubmitFeedbackPayload {
  session_uuid: string;
  submitted_at: string;
}

export interface FeedbackStateResponse {
  session_uuid: string;
  submitted: boolean;
  feedback: FeedbackResponse | null;
}

export interface AssessmentReportResponse {
  session_uuid: string;
  status: string;
  report: ReportOutput;
}

File: backend/app/schemas/session.py
from datetime import datetime
from typing import Any, Literal
from uuid import UUID, uuid4

from pydantic import BaseModel, Field


OccupationCategory = Literal[
    "学生",
    "教育培训",
    "医疗健康",
    "互联网/信息技术",
    "工程/制造/建筑",
    "商业/金融/管理",
    "政府/公共服务",
    "科研/法律/专业服务",
    "文化/传媒/创意",
    "零售/餐饮/生活服务",
    "自由职业/个体经营",
    "待业/退休/其他",
]
CONSENT_VERSION = "critical_thinking_assessment_consent_v1"


class CreateSessionRequest(BaseModel):
    nickname: str = Field(min_length=1, max_length=64)
    occupation_category: OccupationCategory
    occupation: str = Field(min_length=2, max_length=64)
    info_collect_method: str = "ai_dialogue"
    assessment_mode: str = "mock"
    consent_accepted: Literal[True]
    consent_version: Literal["critical_thinking_assessment_consent_v1"]


class StageSummary(BaseModel):
    stage_code: str
    title: str
    stage_order: int
    context: str
    main_question: str
    max_followups: int


class ScenarioSummary(BaseModel):
    scenario_code: str
    title: str
    background: str
    estimated_minutes: int
    version: str
    source_type: str = "seeded"


class OnboardingState(BaseModel):
    question_count: int = 0
    max_questions: int = 3
    completed: bool = False


class ScenarioPreparationState(BaseModel):
    status: str = "queued"
    cache_hit: bool = False
    fallback_used: bool = False
    message: str | None = None


class DialogueTurnItem(BaseModel):
    turn_index: int
    speaker: str
    content: str
    content_type: str
    created_at: datetime
    analysis: dict[str, Any] | None = None


class StageProgressItem(BaseModel):
    stage_code: str
    title: str
    stage_order: int
    status: str
    max_followups: int
    used_followups: int
    used_clarifications: int
    can_skip: bool
    skipped: bool
    released_dynamic_info_count: int
    estimated_minutes: int
    evidence_coverage: dict[str, str] = Field(default_factory=dict)
    missing_evidence: list[str] = Field(default_factory=list)
    waiting_for_stage_choice: bool = False


class SessionProgress(BaseModel):
    total_stages: int
    current_stage_order: int | None
    estimated_minutes: int
    elapsed_seconds: int | None
    stages: list[StageProgressItem] = Field(default_factory=list)


class InterviewProgress(BaseModel):
    formal_answer_count: int = Field(ge=0, le=12)
    target_min_answers: int = 9
    target_max_answers: int = 12
    percent: int = Field(ge=0, le=100)
    estimated_remaining_minutes: int = Field(ge=0)
    elapsed_seconds: int = Field(default=0, ge=0)


class SessionResponse(BaseModel):
    session_uuid: str
    status: str
    flow_version: Literal[
        "legacy_v2", "progressive_v3", "progressive_v3_2", "progressive_v3_3"
    ] = "legacy_v2"
    interviewer_style_version: str = "baseline_v1"
    participant_nickname: str
    scenario: ScenarioSummary
    current_stage: StageSummary | None
    turns: list[DialogueTurnItem] = Field(default_factory=list)
    progress: SessionProgress | None = None
    interview_progress: InterviewProgress | None = None
    language_mode: str = "standard"
    phase: Literal[
        "onboarding", "scenario_preparing", "opening_pending", "assessment", "completed"
    ] = (
        "assessment"
    )
    onboarding: OnboardingState | None = None
    scenario_preparation: ScenarioPreparationState | None = None


class ProfileTurnRequest(BaseModel):
    content: str = Field(min_length=1, max_length=1000)


class PreparationResponse(BaseModel):
    session_uuid: str
    phase: Literal[
        "onboarding", "scenario_preparing", "opening_pending", "assessment", "completed"
    ]
    onboarding: OnboardingState
    scenario_preparation: ScenarioPreparationState
    assessment_ready: bool


class SubmitTurnRequest(BaseModel):
    content: str = Field(min_length=1)
    content_type: str = "scenario_answer"
    client_turn_id: UUID = Field(default_factory=uuid4)
    answer_duration_ms: int | None = Field(default=None, ge=0, le=86_400_000)


class SubmitTurnResponse(BaseModel):
    session_uuid: str
    saved_turn_index: int
    next_action: str
    message: str
    replayed: bool = False


class FinishSessionResponse(BaseModel):
    session_uuid: str
    status: str
    completed_at: datetime


class SkipStageResponse(BaseModel):
    session_uuid: str
    next_action: str
    message: str


class ContinueStageResponse(BaseModel):
    session_uuid: str
    next_action: str
    message: str


class UpdateLanguageModeRequest(BaseModel):
    language_mode: str = Field(pattern="^(standard|plain)$")


class LanguageModeResponse(BaseModel):
    session_uuid: str
    language_mode: str


class SubmitFeedbackRequest(BaseModel):
    realism_score: int = Field(ge=1, le=5)
    difficulty_score: int = Field(ge=1, le=5)
    naturalness_score: int = Field(ge=1, le=5)
    fatigue_score: int = Field(ge=1, le=5)
    report_trust_score: int = Field(ge=1, le=5)
    overall_satisfaction_score: int = Field(ge=1, le=5)
    open_feedback: str | None = Field(default=None, max_length=2000)


class FeedbackResponse(BaseModel):
    session_uuid: str
    realism_score: int
    difficulty_score: int
    naturalness_score: int
    fatigue_score: int
    report_trust_score: int
    overall_satisfaction_score: int
    open_feedback: str | None = None
    submitted_at: datetime


class FeedbackStateResponse(BaseModel):
    session_uuid: str
    submitted: bool
    feedback: FeedbackResponse | None = None


class ReportResponse(BaseModel):
    session_uuid: str
    status: str
    report: dict[str, Any]


class ReportGenerationResponse(BaseModel):
    session_uuid: str
    status: Literal["ready", "scheduled", "running", "failed"]
